File: example/cartesian_teleop_under_lan.cpp
/**
 * @example cartesian_teleop_under_lan.cpp
 * Run Cartesian-space robot-robot teleoperation under LAN (Local Area Network) connection.
 * @copyright Copyright (C) 2016-2024 Flexiv Ltd. All Rights Reserved.
 * @author Flexiv
 */

#include <flexiv/tdk/cartesian_teleop_lan.hpp>

#include <spdlog/spdlog.h>
#include <getopt.h>
#include <iostream>
#include <thread>

namespace {
const struct option kLongOptions[] = {{"first-sn", required_argument, 0, '1'},
    {"second-sn", required_argument, 0, '2'}, {0, 0, 0, 0}};
/** Shaped Cartesian inertia (translation and orientation) */
constexpr std::array<double, 6> kShapedCartInertia = {60.0, 60.0, 60.0, 20.0, 20.0, 20.0};
/** Stiffness ratio and damping ratio (translation and orientation) */
constexpr std::array<double, 6> kCartStiffnessRatio = {0.1, 0.1, 0.1, 0.005, 0.005, 0.005};
constexpr std::array<double, 6> kCartDampingRatio = {0.6, 0.6, 0.6, 0.6, 0.6, 0.6};
}

void PrintHelp()
{
    // clang-format off
    std::cout << "Usage: ./cartesian_teleop_under_lan [-1 serial_num] [-2 serial_num]" << std::endl;
    std::cout << "  -1  --first-sn    Serial number of the first robot." << std::endl;
    std::cout << "  -2  --second-sn   Serial number of the second robot." << std::endl;
    // clang-format on
}

int main(int argc, char* argv[])
{
    // Parse program arguments
    std::string first_sn, second_sn;
    int opt = 0;
    while ((opt = getopt_long(argc, argv, "1:2:", kLongOptions, nullptr)) != -1) {
        switch (opt) {
            case '1':
                first_sn = std::string(optarg);
                break;
            case '2':
                second_sn = std::string(optarg);
                break;
            default:
                PrintHelp();
                return 1;
        }
    }
    if (first_sn.empty() || second_sn.empty()) {
        PrintHelp();
        return 1;
    }

    try {
        // Create teleop control interface
        flexiv::tdk::CartesianTeleopLAN cart_teleop({{first_sn, second_sn}});

        // Only control 1 pair of robots
        unsigned int robot_pair_idx = 0;

        // Run initialization sequence
        cart_teleop.Init();

        // Sync pose, first robot stays still, second robot moves to its pose
        cart_teleop.SyncPose(robot_pair_idx, {});

        // Enable inertia shaping for all Cartesian axes
        std::array<std::pair<bool, double>, flexiv::tdk::kCartDoF> shaped_cart_inertia;
        for (size_t i = 0; i < flexiv::tdk::kCartDoF; i++) {
            shaped_cart_inertia[i] = {true, kShapedCartInertia[i]};
        }
        cart_teleop.SetInertiaShaping(robot_pair_idx, shaped_cart_inertia);

        // Start control loop
        cart_teleop.Start();

        // Set impedance properties
        cart_teleop.SetCartesianImpedance(robot_pair_idx, kCartStiffnessRatio, kCartDampingRatio);

        // Block until faulted
        bool last_pedal_input = false;
        while (!cart_teleop.any_fault()) {
            // Activate by pedal
            bool pedal_input = cart_teleop.digital_inputs(robot_pair_idx).first[0];
            if (pedal_input != last_pedal_input) {
                cart_teleop.Activate(robot_pair_idx, pedal_input);
                last_pedal_input = pedal_input;
            }
            // Sync null-space posture of the second robot to that of the first
            auto first_robot_q = cart_teleop.robot_states(robot_pair_idx).first.q;
            cart_teleop.SetNullSpacePostures(
                robot_pair_idx, std::pair(first_robot_q, first_robot_q));
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        // Fault occurred, stop the teleoperation
        cart_teleop.Stop();

    } catch (const std::exception& e) {
        spdlog::error(e.what());
        return 1;
    }

    return 0;
}

File: example_py/cartesian_teleop_under_lan.py
#!/usr/bin/env python3
"""
Example usage of Cartesian-space robot-robot teleoperation under LAN(Local Area Network) connection.
Monitors digital input from first robot to activate/deactivate teleoperation.
"""

import sys
import argparse
import time
import spdlog
# pip install flexivtdk
import flexiv_tdk


# Constants
SHAPED_CART_INERTIA = [60.0, 60.0, 60.0, 20.0, 20.0, 20.0]
CART_STIFFNESS_RATIO = [0.1, 0.1, 0.1, 0.005, 0.005, 0.005]
CART_DAMPING_RATIO = [0.6, 0.6, 0.6, 0.6, 0.6, 0.6]

def print_help():
    """Print usage information"""
    help_text = """Usage: python cartesian_teleop_under_lan.py [-1 serial_num] [-2 serial_num]
  -1, --first-sn    Serial number of the first robot.
  -2, --second-sn   Serial number of the second robot."""
    print(help_text)

def main():
    parser = argparse.ArgumentParser(
        description="Cartesian-space robot-robot teleoperation under LAN",
        add_help=False
    )
    
    parser.add_argument('-1', '--first-sn', required=True,
                       help='Serial number of the first robot')
    parser.add_argument('-2', '--second-sn', required=True,
                       help='Serial number of the second robot')
    
    # If no arguments provided, show help
    if len(sys.argv) == 1:
        print_help()
        sys.exit(1)
    
    try:
        args = parser.parse_args()
    except SystemExit:
        print_help()
        sys.exit(1)
    
    # Validate required arguments
    if not args.first_sn or not args.second_sn:
        print_help()
        sys.exit(1)
    
    try:
        # Create teleop control interface
        cart_teleop = flexiv.tdk.CartesianTeleopLAN([args.first_sn, args.second_sn])

        # Only control 1 pair of robots
        robot_pair_idx = 0

        # Run initialization sequence
        cart_teleop.Init()

        # Sync pose, first robot stays still, second robot moves to its tcp pose
        first_robot_state = cart_teleop.robot_states(robot_pair_idx).first
        cart_teleop.SyncPose(robot_pair_idx, first_robot_state.tcp_pose)

        # Enable inertia shaping for all Cartesian axes
        shaped_cart_inertia = []
        for i in range(flexiv.tdk.kCartDoF): 
            shaped_cart_inertia.append((True, SHAPED_CART_INERTIA[i]))
        
        cart_teleop.SetInertiaShaping(robot_pair_idx, shaped_cart_inertia)

        # Start control loop
        cart_teleop.Start()

        # Set impedance properties
        cart_teleop.SetCartesianImpedance(
            robot_pair_idx, 
            CART_STIFFNESS_RATIO, 
            CART_DAMPING_RATIO
        )

        # Block until faulted
        last_pedal_input = False
        while not cart_teleop.any_fault():
            # Activate by pedal
            digital_inputs = cart_teleop.digital_inputs(robot_pair_idx).first
            # First digital input of the first robot is used as pedal input
            pedal_input = digital_inputs[0]
            if pedal_input != last_pedal_input:
                cart_teleop.Activate(robot_pair_idx, pedal_input)
                last_pedal_input = pedal_input
            
            # Sync null-space posture of the second robot to that of the first
            first_robot_q = cart_teleop.robot_states(robot_pair_idx).first.q
            # Create a pair with the same joint positions for both robots
            null_space_postures = (first_robot_q, first_robot_q)
            cart_teleop.SetNullSpacePostures(robot_pair_idx, null_space_postures)
            
            time.sleep(0.1)

        # Fault occurred, stop the teleoperation
        cart_teleop.Stop()

    except Exception as e:
        spdlog.error(str(e))
        sys.exit(1)

if __name__ == "__main__":
    main()